### PR TITLE
undo move

### PR DESCRIPTION
**Pull request by dominik1102**
@sebastian-mikucki @tbradlo @rmycek @Konrad-Rys

### CODE REVIEW

**Comment by dominik1102** [comment]
I'm finishing third beer so not so sure ... ;)

File: src/main/java/com/guysfromusa/carsgame/game_state/GameStateTracker.java [hunk 1]
```diff
@@ -19,21 +28,36 @@ public void addNewGame(String gameId){
...
         return this.gameStateMap.get(gameId);
     }
+
+    public List<Movement> getNCarsMovementHistory(String gameId, String carName, int numberOfStepBack){
+        GameState gameState = this.gameStateMap.get(gameId);
+        List<Movement> carsMovement = emptyList();
+        if(gameState != null){
+           carsMovement = gameState.getCarsMovement(carName);
+        }
+        return  carsMovement.isEmpty() ? carsMovement :
```
Comment (tbradlo) [inline src/main/java/com/guysfromusa/carsgame/game_state/GameStateTracker.java]: we switched to Collection due to lack of efficient concurrent List, please try to implement this with iterator

File: src/main/java/com/guysfromusa/carsgame/services/UndoMovementPreparerService.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+    private List<Movement> inverseMovement(List<Movement> movements) {
+        List<Movement> backPath = new ArrayList<>();
+
+        for (Movement movement : movements) {
+            Movement.Operation operation = movement.getOperation();
+            Match(operation).of(
+                    Case($(LEFT), run -> backPath.add(newMovement(RIGHT))),
+                    Case($(RIGHT), run -> backPath.add(newMovement(LEFT))),
+                    Case($(FORWARD), run -> backPath.addAll(asList(newMovement(LEFT), newMovement(LEFT), newMovement(FORWARD)))),
```
[the remark this answers is not in the document]
Comment (tbradlo) [reply src/main/java/com/guysfromusa/carsgame/services/UndoMovementPreparerService.java]: isnt it better (and x times faster) to use switch for Enums?

File: src/main/java/com/guysfromusa/carsgame/services/UndoMovementService.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        this.undoMovementPreparerService = notNull(undoMovementPreparerService);
+    }
+
+    public List<String> doNMoveBack(String gameId, String carName, int numberOfStepBack) throws InterruptedException, ExecutionException {
+        List<Movement> movements = undoMovementPreparerService.prepareBackPath(gameId, carName, numberOfStepBack);
+        List<String> result = new ArrayList<>();
+        Message m = new Message();
+        m.setCarName(carName);
+        m.setGameName(gameId);
+        ScheduledExecutorService scheduler = newSingleThreadScheduledExecutor();
```
Comment (tbradlo) [inline src/main/java/com/guysfromusa/carsgame/services/UndoMovementService.java]: do you destory scheduler? Lets keep it as memeber variable of CommandProducer

File: src/main/java/com/guysfromusa/carsgame/services/UndoMovementService.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+            Future<String> schedule = scheduler.schedule(task, 1, TimeUnit.SECONDS);
+            result.add(schedule.get());
+        }
+        return result;
+    }
+
+    private Callable<String> createTask(Message m) {
+        return () -> {
+            try {
+                gameController.handle(m);
```
Comment (tbradlo) [inline src/main/java/com/guysfromusa/carsgame/services/UndoMovementService.java]: OrderProducer.scheduleMove() or sth similiar

File: src/main/java/com/guysfromusa/carsgame/game_state/GameStateTracker.java (hunk 1 above)
Comment (tbradlo) [reply src/main/java/com/guysfromusa/carsgame/game_state/GameStateTracker.java]: you can switch to COpyOnWriteArrayList

**Pull request merged**